server/service/global_schedule.go

```diff
@@ -0,0 +1,178 @@
+}
+
+type getGlobalScheduleResponse struct {
+	GlobalSchedule []*fleet.ScheduledQuery `json:"global_schedule"`
+	Err            error                   `json:"error,omitempty"`
+}
+
+func (r getGlobalScheduleResponse) error() error { return r.Err }
+
+func getGlobalScheduleEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (interface{}, error) {
+	req := request.(*getGlobalScheduleRequest)
+
+	gp, err := svc.GetGlobalScheduledQueries(ctx, req.ListOptions)
+	if err != nil {
+		return getGlobalScheduleResponse{Err: err}, nil
+	}
+
+	return getGlobalScheduleResponse{
+		GlobalSchedule: gp,
+	}, nil
+}
+
+func (svc *Service) GetGlobalScheduledQueries(ctx context.Context, opts fleet.ListOptions) ([]*fleet.ScheduledQuery, error) {
+	if err := svc.authz.Authorize(ctx, &fleet.Pack{}, fleet.ActionRead); err != nil {
+		return nil, err
+	}
+
+	gp, err := svc.ds.EnsureGlobalPack(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return svc.ds.ListScheduledQueriesInPack(ctx, gp.ID, opts)
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Global Schedule Query
+////////////////////////////////////////////////////////////////////////////////
+
+type globalScheduleQueryRequest struct {
+	QueryID  uint    `json:"query_id"`
+	Interval uint    `json:"interval"`
+	Snapshot *bool   `json:"snapshot"`
+	Removed  *bool   `json:"removed"`
+	Platform *string `json:"platform"`
+	Version  *string `json:"version"`
+	Shard    *uint   `json:"shard"`
+}
+
+type globalScheduleQueryResponse struct {
+	Scheduled *fleet.ScheduledQuery `json:"scheduled,omitempty"`
+	Err       error                 `json:"error,omitempty"`
+}
+
+func (r globalScheduleQueryResponse) error() error { return r.Err }
+
+func globalScheduleQueryEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (interface{}, error) {
+	req := request.(*globalScheduleQueryRequest)
+
+	scheduled, err := svc.GlobalScheduleQuery(ctx, &fleet.ScheduledQuery{
+		QueryID:  req.QueryID,
+		Interval: req.Interval,
+		Snapshot: req.Snapshot,
+		Removed:  req.Removed,
+		Platform: req.Platform,
+		Version:  req.Version,
+		Shard:    req.Shard,
+	})
+	if err != nil {
+		return globalScheduleQueryResponse{Err: err}, nil
+	}
+	return globalScheduleQueryResponse{Scheduled: scheduled}, nil
+}
+
+func (svc *Service) GlobalScheduleQuery(ctx context.Context, sq *fleet.ScheduledQuery) (*fleet.ScheduledQuery, error) {
+	if err := svc.authz.Authorize(ctx, &fleet.Pack{}, fleet.ActionRead); err != nil {
+		return nil, err
+	}
+
+	gp, err := svc.ds.EnsureGlobalPack(ctx)
+	if err != nil {
+		return nil, err
+	}
+	sq.PackID = gp.ID
+
+	return svc.ScheduleQuery(ctx, sq)
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Modify Global Schedule
+////////////////////////////////////////////////////////////////////////////////
+
+type modifyGlobalScheduleRequest struct {
+	ID uint `json:"-" url:"id"`
```

Comment:
I think it's best to have it as we do json-decode the body into that struct (for the payload part, other field). It makes sure this field is left untouched if the request body happened to have an ID field in addition to the expected fields (even though I believe we JSON-decode first, and then apply the URL variables, but should that order ever change it would still be ok).

Reply:
Right. The reason I commented on this was mostly because other IDs are not like this. We might want to see if we can make it so adding `json:"-"` is not needed if you define a `url` or `query` tag.

Reply:
I guess we can set those fields to their zero value after decoding the JSON body so that it nullifies the effect should the JSON set them (the json decoder will still consider those fields and set them if they match a key). It doesn't look like we can dynamically set a struct tag via reflection.

Reply:
Note that I haven't forgotten about this comment! I will address it in a future PR, as a bunch of other PRs have been cut off of this one already.